README.md
# finalhandler

[![NPM Version][npm-image]][npm-url]
[![NPM Downloads][downloads-image]][downloads-url]
[![Node.js Version][node-image]][node-url]
[![Build Status][travis-image]][travis-url]
[![Test Coverage][coveralls-image]][coveralls-url]

Node.js function to invoke as the final step to respond to HTTP request.

## Installation

```sh
$ npm install finalhandler
```

## API

```js
var finalhandler = require('finalhandler')
```

### finalhandler(req, res, [options])

Returns function to be invoked as the final step for the given `req` and `res`.
This function is to be invoked as `fn(err)`. If `err` is falsy, the handler will
write out a 404 response to the `res`. If it is truthy, an error response will
be written out to the `res`, and `res.statusCode` is set from `err.status` if the
value is 4xx or 5xx.

The final handler will also unpipe anything from `req` when it is invoked.

#### options.message

Determine if the error message should be displayed. By default this is false unless
`stacktrace` is enabled. When true, the default behavior will be used to display
the message, which is to use `err.message` if there is a valid `err.status` property.

A function can also be provided to customize the message. The `err` value and outgoing
`status` code are passed and a message string is expected to be returned. If a falsy
value is returned, the response will act as if this option was the default.

```js
var done = finalhandler(req, res, {message: function (err, status) {
  // only display message when err.public is true
  return err.public ? err.message : undefined
}})
```

#### options.onerror

Provide a function to be called with the `err` when it exists. Can be used for
writing errors to a central location without excessive function generation. Called
as `onerror(err, req, res)`.

#### options.stacktrace

Specify if the stack trace of the error should be included in the response. By
default, this is false, but can be enabled when necessary (like in development).
It is not recommend to enable this on a production deployment

## Examples

### always 404

```js
var finalhandler = require('finalhandler')
var http = require('http')

var server = http.createServer(function (req, res) {
  var done = finalhandler(req, res)
  done()
})

server.listen(3000)
```

### perform simple action

```js
var finalhandler = require('finalhandler')
var fs = require('fs')
var http = require('http')

var server = http.createServer(function (req, res) {
  var done = finalhandler(req, res)

  fs.readFile('index.html', function (err, buf) {
    if (err) return done(err)
    res.setHeader('Content-Type', 'text/html')
    res.end(buf)
  })
})

server.listen(3000)
```

### use with middleware-style functions

```js
var finalhandler = require('finalhandler')
var http = require('http')
var serveStatic = require('serve-static')

var serve = serveStatic('public')

var server = http.createServer(function (req, res) {
  var done = finalhandler(req, res)
  serve(req, res, done)
})

server.listen(3000)
```

### display err.message to users

```js
var finalhandler = require('finalhandler')
var http = require('http')

var server = http.createServer(function (req, res) {
  var done = finalhandler(req, res, {message: true})
  var err = new Error('Please try again later')
  err.status = 503
  done(err)
})

server.listen(3000)
```

### display error stack trace for development debugging

```js
var finalhandler = require('finalhandler')
var http = require('http')

var server = http.createServer(function (req, res) {
  var done = finalhandler(req, res, {stacktrace: true})
  var err = new Error('oops')
  done(err)
})

server.listen(3000)
```

### keep log of all errors

```js
var finalhandler = require('finalhandler')
var fs = require('fs')
var http = require('http')

var server = http.createServer(function (req, res) {
  var done = finalhandler(req, res, {onerror: logerror})

  fs.readFile('index.html', function (err, buf) {
    if (err) return done(err)
    res.setHeader('Content-Type', 'text/html')
    res.end(buf)
  })
})

server.listen(3000)

function logerror(err) {
  console.error(err.stack || err.toString())
}
```

## License

[MIT](LICENSE)

[npm-image]: https://img.shields.io/npm/v/finalhandler.svg?style=flat
[npm-url]: https://npmjs.org/package/finalhandler
[node-image]: http://img.shields.io/badge/node.js-%3E%3D_0.8-brightgreen.svg?style=flat
[node-url]: http://nodejs.org/download/
[travis-image]: https://img.shields.io/travis/pillarjs/finalhandler.svg?style=flat
[travis-url]: https://travis-ci.org/pillarjs/finalhandler
[coveralls-image]: https://img.shields.io/coveralls/pillarjs/finalhandler.svg?style=flat
[coveralls-url]: https://coveralls.io/r/pillarjs/finalhandler?branch=master
[downloads-image]: http://img.shields.io/npm/dm/finalhandler.svg?style=flat
[downloads-url]: https://npmjs.org/package/finalhandler


HISTORY.md
unreleased
==========

  * Add `message` option
  * Add `stacktrace` option
  * Add `text/plain` fallback response
  * Ignore `err.status` with non-error code
  * Remove `env` option; use `stacktrace` option instead
  * Send complete HTML document

0.2.0 / 2014-09-03
==================

  * Set `X-Content-Type-Options: nosniff` header
  * deps: debug@~2.0.0

0.1.0 / 2014-07-16
==================

  * Respond after request fully read
    - prevents hung responses and socket hang ups
  * deps: debug@1.0.4

0.0.3 / 2014-07-11
==================

  * deps: debug@1.0.3
    - Add support for multiple wildcards in namespaces

0.0.2 / 2014-06-19
==================

  * Handle invalid status codes

0.0.1 / 2014-06-05
==================

  * deps: debug@1.0.2

0.0.0 / 2014-06-05
==================

  * Extracted from connect/express


test/test.js

var finalhandler = require('..')
var http = require('http')
var request = require('supertest')
var should = require('should')
var stream = require('readable-stream')
var util = require('util')

describe('finalhandler(req, res)', function () {
  describe('status code', function () {
    it('should 404 on no error', function (done) {
      var server = createServer()
      request(server)
      .get('/')
      .expect(404, done)
    })

    it('should 500 on error', function (done) {
      var server = createServer(new Error())
      request(server)
      .get('/')
      .expect(500, done)
    })

    it('should use err.status', function (done) {
      var err = new Error()
      err.status = 400
      var server = createServer(err)
      request(server)
      .get('/')
      .expect(400, done)
    })

    it('should ignore non-error err.status code', function (done) {
      var err = new Error()
      err.status = 201
      var server = createServer(err)
      request(server)
      .get('/')
      .expect(500, done)
    })

    it('should ignore weird err.status', function (done) {
      var err = new Error()
      err.status = 'oh no'
      var server = createServer(err)
      request(server)
      .get('/')
      .expect(500, done)
    })
  })

  describe('404 response', function () {
    it('should include method and path', function (done) {
      var server = createServer()
      request(server)
      .get('/foo')
      .expect(404, /Cannot GET \/foo/, done)
    })

    it('should handle HEAD', function (done) {
      var server = createServer()
      request(server)
      .head('/foo')
      .expect(404, '', done)
    })

    it('should include security header', function (done) {
      var server = createServer()
      request(server)
      .get('/foo')
      .expect('X-Content-Type-Options', 'nosniff')
      .expect(404, done)
    })

    it('should not hang/error if there is a request body', function (done) {
      var buf = new Buffer(1024 * 16)
      var server = createServer()
      var test = request(server).post('/foo')
      buf.fill('.')
      test.write(buf)
      test.write(buf)
      test.write(buf)
      test.expect(404, done)
    })

    describe('when HTML acceptable', function () {
      it('should respond with HTML', function (done) {
        var server = createServer()
        request(server)
        .get('/foo')
        .set('Accept', 'text/html')
        .expect('Content-Type', 'text/html; charset=utf-8')
        .expect(404, /<html/, done)
      })
    })

    describe('when HTML not acceptable', function () {
      it('should respond with plain text', function (done) {
        var server = createServer()
        request(server)
        .get('/foo')
        .set('Accept', 'application/x-bogus')
        .expect('Content-Type', 'text/plain; charset=utf-8')
        .expect(404, 'Cannot GET /foo\n', done)
      })
    })
  })

  describe('error response', function () {
    it('should not include stack trace', function (done) {
      var server = createServer(new Error('boom!'))
      request(server)
      .get('/foo')
      .expect(500, /Internal Server Error/, function (err, res) {
        if (err) return done(err)
        should(res.text).not.match(/boom!/)
        done()
      })
    })

    it('should handle HEAD', function (done) {
      var server = createServer(new Error('boom!'))
      request(server)
      .head('/foo')
      .expect(500, '', done)
    })

    it('should include security header', function (done) {
      var server = createServer(new Error('boom!'))
      request(server)
      .get('/foo')
      .expect('X-Content-Type-Options', 'nosniff')
      .expect(500, done)
    })

    describe('when there is a request body', function () {
      it('should not hang/error when unread', function (done) {
        var buf = new Buffer(1024 * 16)
        var server = createServer(new Error('boom!'))
        var test = request(server).post('/foo')
        buf.fill('.')
        test.write(buf)
        test.write(buf)
        test.write(buf)
        test.expect(500, done)
      })

      it('should not hang/error when actively piped', function (done) {
        var buf = new Buffer(1024 * 16)
        var server = createServer(function (req, res, next) {
          req.pipe(stream)
          process.nextTick(function () {
            next(new Error('boom!'))
          })
        })
        var stream = createSlowWriteStream()
        var test = request(server).post('/foo')
        buf.fill('.')
        test.write(buf)
        test.write(buf)
        test.write(buf)
        test.expect(500, done)
      })

      it('should not hang/error when read', function (done) {
        var buf = new Buffer(1024 * 16)
        var server = createServer(function (req, res, next) {
          // read off the request
          req.once('end', function () {
            next(new Error('boom!'))
          })
          req.resume()
        })
        var test = request(server).post('/foo')
        buf.fill('.')
        test.write(buf)
        test.write(buf)
        test.write(buf)
        test.expect(500, done)
      })
    })

    describe('when HTML acceptable', function () {
      it('should respond with HTML', function (done) {
        var server = createServer(new Error('boom!'))
        request(server)
        .get('/foo')
        .set('Accept', 'text/html')
        .expect('Content-Type', 'text/html; charset=utf-8')
        .expect(500, /<html/, done)
      })
    })

    describe('when HTML not acceptable', function () {
      it('should respond with plain text', function (done) {
        var server = createServer(new Error('boom!'))
        request(server)
        .get('/foo')
        .set('Accept', 'application/x-bogus')
        .expect('Content-Type', 'text/plain; charset=utf-8')
        .expect(500, 'Internal Server Error\n', done)
      })
    })

    describe('when res.statusCode set', function () {
      it('should keep when > 400', function (done) {
        var server = http.createServer(function (req, res) {
          var done = finalhandler(req, res)
          res.statusCode = 503
          done(new Error('oops'))
        })

        request(server)
        .get('/foo')
        .expect(503, done)
      })

      it('should override with err.status', function (done) {
        var server = http.createServer(function (req, res) {
          var done = finalhandler(req, res)
          var err = new Error('oops')
          res.statusCode = 503
          err.status = 414
          done(err)
        })

        request(server)
        .get('/foo')
        .expect(414, done)
      })
    })

    describe('when res.statusCode undefined', function () {
      it('should set to 500', function (done) {
        var server = http.createServer(function (req, res) {
          var done = finalhandler(req, res)
          res.statusCode = undefined
          done(new Error('oops'))
        })

        request(server)
        .get('/foo')
        .expect(500, done)
      })
    })
  })

  describe('request started', function () {
    it('should not respond', function (done) {
      var server = http.createServer(function (req, res) {
        var done = finalhandler(req, res)
        res.statusCode = 301
        res.write('0')
        process.nextTick(done)
      })

      request(server)
      .get('/foo')
      .expect(301, '0', done)
    })
  })

  describe('message', function () {
    it('should reject string', function () {
      finalhandler.bind(null, {}, {}, {message: 'wat'}).should.throw(/option message/)
    })

    it('should display error message', function (done) {
      var err = new Error('boom!')
      err.status = 500
      var server = createServer(err, {message: true})
      request(server)
      .get('/foo')
      .expect(500, /boom!/, done)
    })

    it('should not display error message missing stack property', function (done) {
      var err = new Error('boom!')
      var server = createServer(err, {message: true})
      request(server)
      .get('/foo')
      .expect(500, /Internal Server Error/, function (err, res) {
        if (err) return done(err)
        should(res.text).not.match(/boom!/)
        done()
      })
    })

    it('should not display error message for bad status property', function (done) {
      var err = new Error('boom!')
      err.status = 'oh no'
      var server = createServer(err, {message: true})
      request(server)
      .get('/foo')
      .expect(500, /Internal Server Error/, function (err, res) {
        if (err) return done(err)
        should(res.text).not.match(/boom!/)
        done()
      })
    })

    it('should escape message for HTML response', function (done) {
      var err = new Error('<boom>!')
      err.status = 500
      var server = createServer(err, {message: true})
      request(server)
      .get('/foo')
      .set('Accept', 'text/html')
      .expect(500, /&lt;boom&gt;!/, done)
    })

    describe('when function', function () {
      it('should use custom function for message', function (done) {
        var err = new Error('boom!')
        var server = createServer(err, {message: function (err, status) {
          return 'custom ' + status + ' ' + err.message
        }})

        request(server)
        .get('/foo')
        .expect(500, /custom 500 boom!/, done)
      })

      it('should provide fallback for custom function', function (done) {
        var err = new Error('boom!')
        var server = createServer(err, {message: function (err) {
          return undefined
        }})

        request(server)
        .get('/foo')
        .expect(500, /Internal Server Error/, function (err, res) {
          if (err) return done(err)
          should(res.text).not.match(/boom!/)
          done()
        })
      })

      it('should escape message for HTML response', function (done) {
        var err = new Error('<boom>!')
        var server = createServer(err, {message: function (err) {
          return 'custom ' + err.message
        }})

        request(server)
        .get('/foo')
        .set('Accept', 'text/html')
        .expect(500, /custom &lt;boom&gt;!/, done)
      })
    })
  })

  describe('onerror', function () {
    it('should be invoked when error', function (done) {
      var err = new Error('boom!')
      var error
      var log = function (e) { error = e }
      var server = createServer(err, {onerror: log})

      request(server)
      .get('/')
      .end(function () {
        should(error).equal(err)
        done()
      })
    })
  })

  describe('stacktrace', function () {
    it('should include error stack', function (done) {
      var server = createServer(new Error('boom!'), {stacktrace: true})
      request(server)
      .get('/foo')
      .expect(500, /Error: boom!.*at.*:[0-9]+:[0-9]+/, done)
    })

    it('should escape error stack for HTML response', function (done) {
      var server = createServer(new Error('boom!'), {stacktrace: true})
      request(server)
      .get('/foo')
      .set('Accept', 'text/html')
      .expect(500, /Error: boom!<br> &nbsp; &nbsp;at/, done)
    })

    it('should not escape error stack for plain text response', function (done) {
      var server = createServer(new Error('boom!'), {stacktrace: true})
      request(server)
      .get('/foo')
      .set('Accept', 'application/x-bogus')
      .expect('Content-Type', 'text/plain; charset=utf-8')
      .expect(500, /Error: boom!\n    at/, done)
    })

    it('should handle non-error-objects', function (done) {
      var server = createServer('lame string', {stacktrace: true})
      request(server)
      .get('/foo')
      .set('Accept', 'text/html')
      .expect(500, /lame string/, done)
    })

    describe('when message set', function () {
      it('should use custom function for message', function (done) {
        var err = new Error('boom!')
        var server = createServer(err, {stacktrace: true, message: function (err, status) {
          return 'custom ' + status + ' ' + err.message
        }})

        request(server)
        .get('/foo')
        .expect(500, /Error: custom 500 boom!.*at.*:[0-9]+:[0-9]+/, done)
      })

      it('should provide fallback for custom function', function (done) {
        var err = new Error('boom!')
        var server = createServer(err, {stacktrace: true, message: function (err, status) {
          return undefined
        }})

        request(server)
        .get('/foo')
        .expect(500, /Error: boom!.*at.*:[0-9]+:[0-9]+/, done)
      })

      it('should handle non-error-objects', function (done) {
        var err = 'lame string'
        var server = createServer(err, {stacktrace: true, message: true})

        request(server)
        .get('/foo')
        .expect(500, /lame string/, done)
      })
    })
  })
})

function createServer(err, opts) {
  return http.createServer(function (req, res) {
    var done = finalhandler(req, res, opts)

    if (typeof err === 'function') {
      err(req, res, done)
      return
    }

    done(err)
  })
}

function createSlowWriteStream() {
  return new SlowWriteStream()
}

function SlowWriteStream() {
  stream.Writable.call(this)
}

util.inherits(SlowWriteStream, stream.Writable)

SlowWriteStream.prototype._write = function _write(chunk, encoding, callback) {
  setTimeout(callback, 1000)
}


index.js
/*!
 * finalhandler
 * Copyright(c) 2014 Douglas Christopher Wilson
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var accepts = require('accepts')
var debug = require('debug')('finalhandler')
var escapeHtml = require('escape-html')
var http = require('http')

/**
 * Variables.
 */

/* istanbul ignore next */
var defer = typeof setImmediate === 'function'
  ? setImmediate
  : function(fn){ process.nextTick(fn.bind.apply(fn, arguments)) }

/**
 * Module exports.
 */

module.exports = finalhandler

/**
 * Final handler:
 *
 * @param {Request} req
 * @param {Response} res
 * @param {Object} [options]
 * @return {Function}
 * @api public
 */

function finalhandler(req, res, options) {
  options = options || {}

  // get message option
  var message = options.message === true
    ? getDefaultErrorMessage
    : options.message || false

  if (typeof message !== 'boolean' && typeof message !== 'function') {
    throw new TypeError('option message must be boolean or function')
  }

  // get error callback
  var onerror = options.onerror

  // get stack trace option
  var stacktrace = options.stacktrace || false;

  return function (err) {
    var body
    var constructBody
    var msg
    var status = res.statusCode

    // unhandled error
    if (err) {
      // default status code to 500
      if (!status || status < 400) {
        status = 500
      }

      // respect err.status
      if (err.status >= 400 && err.status < 600) {
        status = err.status
      }

      // build a stack trace or normal message
      msg = stacktrace
        ? getErrorStack(err, status, message)
        : getErrorMessage(err, status, message)
    } else {
      status = 404
      msg = 'Cannot ' + req.method + ' ' + (req.originalUrl || req.url)
    }

    debug('default %s', status)

    // schedule onerror callback
    if (err && onerror) {
      defer(onerror, err, req, res)
    }

    // cannot actually respond
    if (res._header) {
      return req.socket.destroy()
    }

    // negotiate
    var accept = accepts(req)
    var type = accept.types('html', 'text')

    // construct body
    switch (type) {
      case 'html':
        constructBody = constructHtmlBody
        break
      default:
        // default to plain text
        constructBody = constructTextBody
        break
    }

    // construct body
    body = constructBody(status, msg)

    // send response
    send(req, res, status, body)
  }
}

/**
 * Get HTML body string
 *
 * @param {number} status
 * @param {string} message
 * @return {Buffer}
 * @api private
 */

function constructHtmlBody(status, message) {
  var msg = escapeHtml(message)
    .replace(/\n/g, '<br>')
    .replace(/  /g, ' &nbsp;')

  var html = '<!doctype html>\n'
    + '<html lang=en>\n'
    + '<head>\n'
    + '<meta charset=utf-8>\n'
    + '<title>' + escapeHtml(http.STATUS_CODES[status]) + '</title>\n'
    + '</head>\n'
    + '<body>\n'
    + msg + '\n'
    + '</body>\n'

  var body = new Buffer(html, 'utf8')

  body.type = 'text/html; charset=utf-8'

  return body
}

/**
 * Get plain text body string
 *
 * @param {number} status
 * @param {string} message
 * @return {Buffer}
 * @api private
 */

function constructTextBody(status, message) {
  var msg = message + '\n'
  var body = new Buffer(msg, 'utf8')

  body.type = 'text/plain; charset=utf-8'

  return body
}

/**
 * Get message from error
 *
 * @param {object} err
 * @param {number} status
 * @param {function} message
 * @return {string}
 * @api private
 */

function getErrorMessage(err, status, message) {
  var msg

  if (message) {
    msg = message(err, status)
  }

  return msg || http.STATUS_CODES[status]
}

/**
 * Get default message from error
 *
 * @param {object} err
 * @return {string}
 * @api private
 */

function getDefaultErrorMessage(err) {
  return err.status >= 400 && err.status < 600
    ? err.message
    : undefined
}

/**
 * Get stack from error with custom message
 *
 * @param {object} err
 * @param {number} status
 * @param {function} message
 * @return {string}
 * @api private
 */

function getErrorStack(err, status, message) {
  var stack = err.stack || ''

  if (message) {
    var index = stack.indexOf('\n')
    var msg = message(err, status) || err.message || String(err)
    var name = err.name

    // slice implicit message from top of stack
    if (index !== -1) {
      stack = stack.substr(index)
    }

    // prepend name and message to stack
    stack = name
      ? name + ': ' + msg + stack
      : msg + stack
  } else if (!stack) {
    // stringify error when no message generator and no stack
    stack = String(err)
  }

  return stack
}

/**
 * Send response.
 *
 * @param {IncomingMessage} req
 * @param {OutgoingMessage} res
 * @param {number} status
 * @param {Buffer} body
 * @api private
 */

function send(req, res, status, body) {
  function write() {
    res.statusCode = status

    // security header for content sniffing
    res.setHeader('X-Content-Type-Options', 'nosniff')

    // standard headers
    res.setHeader('Content-Type', body.type)
    res.setHeader('Content-Length', body.length)

    if (req.method === 'HEAD') {
      res.end()
      return
    }

    res.end(body, 'utf8')
  }

  if (!req.readable) {
    write()
    return
  }

  // unpipe everything from the request
  unpipe(req)

  // flush the request
  req.once('end', write)
  req.resume()
}

/**
 * Unpipe everything from a stream.
 *
 * @param {Object} stream
 * @api private
 */

/* istanbul ignore next: implementation differs between versions */
function unpipe(stream) {
  if (typeof stream.unpipe === 'function') {
    // new-style
    stream.unpipe()
    return
  }

  // Node.js 0.8 hack
  var listener
  var listeners = stream.listeners('close')

  for (var i = 0; i < listeners.length; i++) {
    listener = listeners[i]

    if (listener.name !== 'cleanup' && listener.name !== 'onclose') {
      continue
    }

    // invoke the listener
    listener.call(stream)
  }
}
